src/charm.py

```diff
@@ -95,6 +143,107 @@ def _pebble_layer(self) -> ops.pebble.LayerDict:
+        if req.status_code != 200:
+            return False
+        body = req.json()
+        return body.get("initialized", False)
+
+    ## Helpers ##
+    def _generate_self_signed_certificates(self) -> None:
+        """Generate self signed certificates and saves them to secrets and the charm."""
+        logger.info("[GoCert] Creating self signed certificates.")
+        if not self._application_bind_address:
+            logger.warning("[GoCert] unit IP not found.")
+            return
+        ca, ca_pk = self._get_ca_certificate()
+        pk = generate_private_key()
+        csr = generate_csr(
+            private_key=pk,
+            subject="GoCert Self Signed Certificate",
+            sans_ip=[self._application_bind_address],
+        )
+        cert = generate_certificate(csr=csr, ca=ca, ca_key=ca_pk)
+
+        self.container.push(f"{WORKLOAD_PATH}/{CONFIG_MOUNT}/ca.pem", ca, make_dirs=True)
+        self.container.push(
+            f"{WORKLOAD_PATH}/{CONFIG_MOUNT}/certificate.pem", cert, make_dirs=True
+        )
+        self.container.push(f"{WORKLOAD_PATH}/{CONFIG_MOUNT}/private_key.pem", pk, make_dirs=True)
+        logger.info("[GoCert] Created self signed certificates.")
+
+    def _self_signed_certificates_generated(self) -> bool:
+        """Check if the workload certificate was generated and was self signed."""
+        try:
+            existing_cert = self.container.pull(f"{WORKLOAD_PATH}/{CONFIG_MOUNT}/certificate.pem")
+        except ops.pebble.PathError:
+            return False
+        try:
+            certificate = x509.load_pem_x509_certificate(existing_cert.read().encode())
+            common_name = certificate.issuer.get_attributes_for_oid(x509.NameOID.COMMON_NAME)[0]
+        except (ValueError, TypeError):
+            return False
+        if common_name.value != SELF_SIGNED_CA_COMMON_NAME:
+            logger.warning(common_name.value)
+            logger.warning(SELF_SIGNED_CA_COMMON_NAME)
+            return False
+        return True
+
+    def _get_ca_certificate(self) -> Tuple[bytes, bytes]:
+        """Get the CA certificate from secrets. Create one if it doesn't exist."""
+        try:
+            secret = self.model.get_secret(label=SELF_SIGNED_CA_SECRET_LABEL)
+            secret_content = secret.get_content(refresh=True)
+            ca_cert = secret_content.get("certificate", "")
+            ca_pk = secret_content.get("private-key", "")
+            content = CertificateSecret(
+                certificate=ca_cert,
+                private_key=ca_pk,
+            )
+        except ops.SecretNotFoundError:
+            pk = generate_private_key()
+            ca = generate_ca(private_key=pk, subject=SELF_SIGNED_CA_COMMON_NAME)
+            content = CertificateSecret(certificate=ca.decode(), private_key=pk.decode())
+            self.app.add_secret(label=SELF_SIGNED_CA_SECRET_LABEL, content=content.to_dict())
+        return content.certificate.encode(), content.private_key.encode()
```

Comment:
I'm not a huge fan that this `get_xyz` can create stuff, it feels like we should either change the function name or separate it in to.

Reply:
The docstring mentions that it will be created but i renamed the function too.